src/sacn/source.c

```diff
@@ -1033,42 +1026,166 @@ int sacn_source_process_manual(void)
+ * work for a universe.
+ *
+ * @param[in, out] netint_lists Lists of interfaces the application wants to use for each universe. Must not be NULL.
+ * Must include all universes of all sources, and nothing more. The status codes are filled in whenever
+ * SacnSourceUniverseNetintList::netints is non-NULL.
+ * @param[in] num_netint_lists The size of netint_lists. Must not be 0.
+ * @return #kEtcPalErrOk: Networking reset successfully.
+ * @return #kEtcPalErrNoNetints: None of the network interfaces provided for a universe were usable by the library.
+ * @return #kEtcPalErrInvalid: Invalid parameter provided.
+ * @return #kEtcPalErrNotInit: Module not initialized.
+ * @return #kEtcPalErrSys: An internal library or system call error occurred.
+ */
+etcpal_error_t sacn_source_reset_networking_per_universe(const SacnSourceUniverseNetintList* netint_lists,
+                                                         size_t num_netint_lists)
+{
+#if SACN_SOURCE_ENABLED
+  etcpal_error_t result = kEtcPalErrOk;
+
+  if (!sacn_initialized())
+    result = kEtcPalErrNotInit;
+
+  if ((netint_lists == NULL) || (num_netint_lists == 0))
+    result = kEtcPalErrInvalid;
+
+  if (sacn_lock())
+  {
+    // Validate netint_lists. It must include all universes of all sources and nothing more.
+    size_t total_num_universes = 0;
+    for (size_t i = 0; (result == kEtcPalErrOk) && (i < get_num_sources()); ++i)
+    {
+      for (size_t j = 0; (result == kEtcPalErrOk) && (j < get_source(i)->num_universes); ++j)
+      {
+        ++total_num_universes;
+
+        bool found = false;
+        for (size_t k = 0; !found && (k < num_netint_lists); ++k)
+        {
+          found = ((get_source(i)->handle == netint_lists[k].handle) &&
+                   (get_source(i)->universes[j].universe_id == netint_lists[k].universe));
+        }
+
+        if (!found)
+          result = kEtcPalErrInvalid;
+      }
+    }
+
+    if (result == kEtcPalErrOk)
+    {
+      if (num_netint_lists != total_num_universes)
+        result = kEtcPalErrInvalid;
+    }
+
+    if (result == kEtcPalErrOk)
+    {
+      sacn_sockets_reset_source();
+
+      for (size_t i = 0; i < get_num_sources(); ++i)
+        get_source(i)->num_netints = 0;  // Clear source netints, will be reconstructed when netints are re-added.
+    }
+
+    for (size_t i = 0; (result == kEtcPalErrOk) && (i < num_netint_lists); ++i)
+    {
+      const SacnSourceUniverseNetintList* netint_list = &netint_lists[i];
+
+      SacnSource* source;
+      SacnSourceUniverse* universe;
+      lookup_source_and_universe(netint_list->handle, netint_list->universe, &source, &universe);
+
+      universe->netints.num_netints = 0;
+      result = sacn_initialize_source_netints(&universe->netints, netint_list->netints, netint_list->num_netints);
+
+      for (size_t j = 0; (result == kEtcPalErrOk) && (j < universe->netints.num_netints); ++j)
+        result = add_to_source_netints(source, &universe->netints.netints[j]);
```

Comment:
As before, we want all valid network interfaces, right?  Wouldn't this break early?

Reply:
sacn_initialize_source_netints initializes universe->netints with only the valid netints. I don't see where this breaks early.

Reply:
It does not.  The misunderstanding was on my side.